src/lsps2/channel_manager.rs

```diff
@@ -1302,36 +1302,60 @@ fn calculate_amount_to_forward_per_htlc(
 mod tests {
 
 	use super::*;
+	use proptest::prelude::*;
 
-	#[test]
-	fn test_calculate_amount_to_forward() {
-		// TODO: Use proptest to generate random allocations
-		let htlcs = vec![
-			InterceptedHTLC {
-				intercept_id: InterceptId([0; 32]),
-				expected_outbound_amount_msat: 1000,
-			},
-			InterceptedHTLC {
-				intercept_id: InterceptId([1; 32]),
-				expected_outbound_amount_msat: 2000,
-			},
-			InterceptedHTLC {
-				intercept_id: InterceptId([2; 32]),
-				expected_outbound_amount_msat: 3000,
-			},
-		];
-
-		let total_amt_to_forward_msat = 5000;
+	const MAX_VALUE_MSAT: u64 = 21_000_000_0000_0000_000;
 
-		let result = calculate_amount_to_forward_per_htlc(&htlcs, total_amt_to_forward_msat);
-
-		assert_eq!(result[0].0, htlcs[0].intercept_id);
-		assert_eq!(result[0].1, 834);
+	fn arb_forward_amounts() -> impl Strategy<Value = (u64, u64, u64, u64)> {
+		(1u64..MAX_VALUE_MSAT, 1u64..MAX_VALUE_MSAT, 1u64..MAX_VALUE_MSAT, 1u64..MAX_VALUE_MSAT)
+			.prop_map(|(a, b, c, d)| {
+				(a, b, c, core::cmp::min(d, a.saturating_add(b).saturating_add(c)))
+			})
+	}
 
-		assert_eq!(result[1].0, htlcs[1].intercept_id);
-		assert_eq!(result[1].1, 1667);
+	proptest! {
+		#[test]
+		fn test_calculate_amount_to_forward((o_0, o_1, o_2, total_amt_to_forward_msat) in arb_forward_amounts()) {
+			let htlcs = vec![
+				InterceptedHTLC {
+					intercept_id: InterceptId([0; 32]),
+					expected_outbound_amount_msat: o_0
+				},
+				InterceptedHTLC {
+					intercept_id: InterceptId([1; 32]),
+					expected_outbound_amount_msat: o_1
+				},
+				InterceptedHTLC {
+					intercept_id: InterceptId([2; 32]),
+					expected_outbound_amount_msat: o_2
+				},
+			];
+
+			let result = calculate_amount_to_forward_per_htlc(&htlcs, total_amt_to_forward_msat);
+			let total_received_msat = o_0 + o_1 + o_2;
+
+			if total_received_msat < total_amt_to_forward_msat {
+				assert_eq!(result.len(), 0);
+			} else {
+				assert_ne!(result.len(), 0);
+				assert_eq!(result[0].0, htlcs[0].intercept_id);
+				assert_eq!(result[1].0, htlcs[1].intercept_id);
+				assert_eq!(result[2].0, htlcs[2].intercept_id);
+				assert!(result[0].1 <= o_0);
+				assert!(result[1].1 <= o_1);
+				assert!(result[2].1 <= o_2);
```

Comment:
probably should also assert that it doesn't equal 0. I think this is an edge case the current implementation will fail.

Reply:
hm but returning 0 should be a bug since you can't forward a 0 value htlc? no? this is one reason I think this method will need to return a Result to actually handle all the edge cases properly